pgsql/pc_editor.c

```diff
@@ -21,32 +21,46 @@ Datum pcpatch_setschema(PG_FUNCTION_ARGS)
 	int32 mode = PG_GETARG_INT32(3);
 	PCSCHEMA *schema = pc_schema_from_pcid(serpa->pcid, fcinfo);
 	PCSCHEMA *new_schema = pc_schema_from_pcid(pcid, fcinfo);
-	PCPATCH *patch = pc_patch_deserialize(serpa, schema);
-	bool reinterpret;
-	float8 defaultvalue;
+	PCPATCH *patch;
+	float8 defaultvalue = 0.0;
 
-	if ( ! patch )
-		PG_RETURN_NULL();
-
-	switch ( mode )
+	if ( mode == 0 )
 	{
-	case 0:
+		bool reinterpret;
+		if ( pc_schema_similar(schema, new_schema) )
+		{
+			// fast path!
+			serpatch = palloc(serpa->size);
+			if ( ! serpatch )
+				PG_RETURN_NULL();
+			memcpy(serpatch, serpa, serpa->size);
+			serpatch->pcid = pcid;
+			PG_RETURN_POINTER(serpatch);
```

Comment:
Not sure, AFAIK, the compression is a property of the serpatch that is used for deserialization : https://github.com/pgpointcloud/pointcloud/blob/master/pgsql/pc_pgsql.c#L984.
Could you design some tests to check that ?